src/server/worker/api_server.go

```diff
@@ -838,6 +877,12 @@ func (a *APIServer) uploadOutput(pachClient *client.APIClient, dir string, tag s
 	if err := putObjsClient.CloseSend(); err != nil {
 		return err
 	}
+	// Serialize datum hashtree
+	b := &bytes.Buffer{}
```

Comment:
Why serialize into a `bytes.Buffer` here rather than just directly into `w` below? Does it help to batch things up? And is that worth the extra memory used for the `bytes.Buffer`?

Reply:
I need to store it in a byte buffer so I can cache it. If I just wrote directly into w, with the current model, I would need to re-serialize it into the cache. I decided it would be better to store a single serialized hashtree in memory rather than serialize every datum hashtree twice. There is probably a better way to do this such as writing to two locations in the serialize function, but I went with this because it is easy to reason about and the extra memory (one serialized datum hashtree) should not be much in most cases.

Reply:
Makes sense, I didn't see you were using the buffer below to put it into the `datumCache`.

I think you could do this pretty easily with `io.MultiWriter`.

Reply:
That looks like exactly what I need. I think I will switch over to that.